### ISSUE TITLE
Obscure error message when field or any sub-field is undefined.

### ISSUE THREAD

**Issue by Xheotris** [rethinkdb/rethinkdb#4445]
It's a small nitpick, but the error:

    ./node_modules/rethinkdb/node_modules/bluebird/js/main/async.js:43
            fn = function () { throw arg; };
                                     ^
    RqlDriverError: Cannot wrap undefined with r.expr().

Is quite confusing if you don't know what to look for. As I was trying to debug, I was looking for something wrong with the query, the callback, or the connection, instead of the data being fed to it, based on the fact that it was complaining about an expression and showing me a function. This threw for me on a query along the lines of `r.table('things').insert(document).run(conn, myFun)` The fact that it was complaining about `r.expr()` was truly mystifying.

I would change the text to make it clear that it's an issue with a data field. Perhaps `Cannot store undefined field.` might work better?

**Comment by mlucy** [comment]
Thanks for the report @Xheotris !

@deontologician, does it seem reasonable to improve the error message here?

**Comment by deontologician** [comment]
I agree its confusing. I'm not sure why we catch this particular problem so late in the processing

**Comment by neumino** [comment]
For what it's worth, rethinkdbdash has been dropping `undefined` fields since its first release and so far no one complained about it

It still throws an error if you try to insert `undefined` (or if a function returns undefined), but if just one field is undefined, it just skips the key. One case where fields mapping to undefined appear is when you (or a library) do:
```
var doc = {
  foo: "bar",
  buzz: (condition) ? "buzz" : undefined
}
```
instead of
```
var doc = {foo: "bar"};
if (condition) {
  doc.buzz = "buzz"
}
```

**Comment by coffeemug** [comment]
We've covered some distance since we first designed this, and we have a little bit more perspective now and understanding of how people use the system. I think throwing on `undefined` when evaluating json document expressions is always too aggressive -- it's clearly not what people want the system to do. I think it would be dramatically better for user experience to just drop the undefined fields (though it's probably not too important to do fast because people don't run into it *that* frequently).

**Comment by mlucy** [comment]
I think I agree; let's just drop them.

**Comment by neumino** [comment]
:+1:

**Comment by jamesgibson14** [comment]
1+ for dropping undefined fields.
